src/Sema.zig

```diff
@@ -26264,8 +26264,16 @@ fn explainWhyTypeIsNotExtern(
             try mod.errNoteNonLazy(src_loc, msg, "enum tag type '{}' is not extern compatible", .{tag_ty.fmt(sema.mod)});
             try sema.explainWhyTypeIsNotExtern(msg, src_loc, tag_ty, position);
         },
-        .Struct => try mod.errNoteNonLazy(src_loc, msg, "only extern structs and ABI sized packed structs are extern compatible", .{}),
-        .Union => try mod.errNoteNonLazy(src_loc, msg, "only extern unions and ABI sized packed unions are extern compatible", .{}),
+        .Struct => switch (ty.containerLayout(mod)) {
+            .Extern => return,
+            .Packed => try mod.errNoteNonLazy(src_loc, msg, "only 0, 8, 16, 32, 64 and 128 bits sized packed structs are extern compatible", .{}),
+            .Auto => try mod.errNoteNonLazy(src_loc, msg, "only empty struct without comptime dependency are extern compatible", .{}),
+        },
+        .Union => switch (ty.containerLayout(mod)) {
+            .Extern => return,
+            .Packed => try mod.errNoteNonLazy(src_loc, msg, "only 0, 8, 16, 32, 64 and 128 bits sized packed unions are extern compatible", .{}),
+            .Auto => try mod.errNoteNonLazy(src_loc, msg, "only empty union without comptime dependency are extern compatible", .{}),
```

Comment:
I think these new errors are much worse. IMO non-extern structs shouldn't be allowed at all and they definitely shouldn't be suggested by the errors.

Reply:
Well, struct{} and union{} are allowed, because of https://github.com/ziglang/zig/issues/16394. The .Extern case is indeed unreachable, but I just kept the same style of switch handling as code above, which return on .Bool/.Float which are also by default extern compatible.

Reply:
I can revert error message, and just add something like this in the end "empty union/structs should not be comptime dependent".

Reply:
I saw that I just don't agree with it and don't think an edge case like this should be mentioned in the error message.

Reply:
Ok, I will fix error message.
Do you have any opinions about my message regarding `void` above? I will appreciate some feedback from a core team member.

Reply:
I'm not sure what `void` you're referencing?

Reply:
My comment above: https://github.com/ziglang/zig/pull/18749#issuecomment-1917595941.

Reply:
Ah, I didn't see it because GitHub sent me to the "files changed" view from the notification. I think all three should be errors but since they aren't void shouldn't be either.

Reply:
Ok, I will probably extract this to issues as bug/proposal